File: leetcode/00692_top-k-frequent-words/692-top-k-frequent-words.py
class Solution:
    def topKFrequent(self, words: List[str], k: int) -> List[str]:
        d = {} # count_dict
        for c in words:
            if c not in d:
                d[c] = 1
            else:
                d[c] += 1
        
        # sorted: frequency
        #sorted_list = [k for k, v in sorted(d.items(), key=lambda item: item[1], reverse=True)]
        sorted_list = [k for k, v in sorted(d.items(), key=cmp_to_key(self.compare))]

        # Top K
        return sorted_list[:k]
    
    def compare(self, x, y):
        # x: post, y: prev
        # It needs to return -1, 0, or 1
        if x[1] == y[1]:
            # lexicographical order: greater
            return 1 if x[0] > y[0] else -1
        else:
            # less
            return 1 if x[1] < y[1] else -1
        

File: leetcode/00692_top-k-frequent-words/692-top-k-frequent-words.cpp
typedef pair<string, int> Node;
class Compare {
 public:
  bool operator()(const Node& a, const Node& b) const {
    if (a.second == b.second) {
        // less
        return (a.first > b.first);
    }
    // greater
    return a.second < b.second;
  }
};

class Solution {
public:
    vector<string> topKFrequent(vector<string>& words, int k) {
        // count
        unordered_map<string, int> count;
        for (const auto& word : words)
            ++count[word];
        
        // priority queue by frequency
        priority_queue<Node, vector<Node>, Compare> q;
        for (const auto& kv: count) {
            q.push(kv);
        }
        
        vector<string> ans;
        while (!q.empty()) {
            ans.push_back(q.top().first);
            q.pop();
            
            if (ans.size() >= k) 
                break;
        }
        
        return ans;
    }
};